core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -325,24 +309,50 @@ sealed trait Cause[+E] extends Product with Serializable { self =>
       case Meta(c, data)    => c.stripFailures.map(Meta(_, data))
     }
 
-  final def succeeded: Boolean = !failed
-
   final def traces: List[ZTrace] =
     self
       .foldLeft(List.empty[ZTrace]) {
         case (z, Traced(_, trace)) => trace :: z
       }
       .reverse
 
-  private def foldLeft[Z](z: Z)(f: PartialFunction[(Z, Cause[E]), Z]): Z =
-    (f.lift(z -> self).getOrElse(z), self) match {
-      case (z, Then(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
-      case (z, Both(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
-      case (z, Traced(cause, _))  => cause.foldLeft(z)(f)
-      case (z, Meta(cause, _))    => cause.foldLeft(z)(f)
+  private def find[Z](f: PartialFunction[Cause[E], Z]): Option[Z] = {
+    @tailrec
+    def loop(cause: Cause[E], stack: List[Cause[E]]): Option[Z] =
+      f.lift(cause) match {
+        case Some(z) => Some(z)
+        case None =>
+          cause match {
```

Comment:
This is `cause.foldLeft(None) { case (_, cause) if f.isDefined(cause) => Some(f(cause)) }`, so find could be written as
```scala
private def find[Z](f: PartialFunction[Cause[E], Z]): Option[Z] = {
  f.lift(self) orElse foldLeft(None){ case (_, cause) if f.isDefined(cause) => Some(f(cause)) }
```

Reply:
I think this would require us to traverse the entire data structure if `f(lift(self)` is `None`, because then we call `foldLeft` which always traverses the entire data structure. The `PartialFunction` in `foldLeft` controls whether a value from a particular cause will be added to the accumulator but can't terminate the traversal early.

Reply:
Ha yeah right, you would need lazyness to get around that. Shame, that is elegant:
```scala
private def find[Z](f: PartialFunction[Cause[E], Z]): Option[Z] = {
  f.lift(self) orElse foldLeft(None){ case (z, cause) if f.isDefinedAt(cause) => z orElse Some(f(cause)) }
```
Out of curiosity, why a `PartialFunction` in `foldLeft`?